### PR TITLE
Train model

### PR DESCRIPTION
**Pull request by bdittes**
- Introduce the 'trainer' module, based on huridocs/classification_utils.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -31,3 +31,16 @@ To run operations from the command line, `./run local --help`.
 To run nose tests, `./run test`.
 
 To run pycodestyle, `./run lint`.
+
```
Comment (Tilana) [inline README.md]: add documentation on how to use ./run train

File: app/trainer.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+train_bp = Blueprint('train_bp', __name__)
+
+# LEARNING_RATE = 2e-3
+LEARNING_RATE = 0.02
+NUM_WARMUP_STEPS = 100
+WARMUP_PROPORTION = 0.0
+DROPOUT = 0.1
+# SHARED_SIZE = 512
+SHARED_SIZE = 0
+BATCH_SIZE = 256
```
Comment (Tilana) [inline app/trainer.py]: The batch size regulates how much of the training data is stored at once in memory. Dependent on the used machine 256 might be already too much.
Let's add this at least to the documentation...

File: app/trainer.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+import app.tasks as tasks
+from app.classifier import Classifier, ClassifierCache
+from app.embedder import MAX_SEQ_LENGTH, Embedder
+
+train_bp = Blueprint('train_bp', __name__)
+
+# LEARNING_RATE = 2e-3
+LEARNING_RATE = 0.02
+NUM_WARMUP_STEPS = 100
+WARMUP_PROPORTION = 0.0
```
Comment (Tilana) [inline app/trainer.py]: unused variable

File: app/trainer.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+import app.tasks as tasks
+from app.classifier import Classifier, ClassifierCache
+from app.embedder import MAX_SEQ_LENGTH, Embedder
+
+train_bp = Blueprint('train_bp', __name__)
+
+# LEARNING_RATE = 2e-3
+LEARNING_RATE = 0.02
+NUM_WARMUP_STEPS = 100
```
Comment (Tilana) [inline app/trainer.py]: If we are not training BERT (which is the default scenario), NUM_WARMUP_STEPS has to be set to 0.

File: app/trainer.py [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+
+def _one_hot_labels(label_list: List[str], all_labels: List[str]) -> List[int]:
+    labels = [0] * len(all_labels)
+    indices = [all_labels.index(label) for label in label_list]
+    for ind in indices:
+        labels[ind] = 1
+    return labels
+
+
+class InputExample(object):
```
Comment (Tilana) [inline app/trainer.py]: trainer class is too long. Please create a new module for InputExample, InputFeatures, inputf_fn_builder, etc.

File: app/trainer.py [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+DEFAULT_BERT = 'https://tfhub.dev/google/bert_uncased_L-12_H-768_A-12/1'
+
+
+class _TrainModel(tasks.TaskProvider):
+
+    def __init__(self, json: Any):
+        super().__init__(json)
+        self.base_classifier_dir = json['base_classifier_dir']
+        self.model = json['model']
+        self.vocab = json['vocab'] if 'vocab' in json else None
```
Comment (Tilana) [inline app/trainer.py]: Vocab is always a property of the bert base model. It shouldn't be possible to load a vocab file with a BERT model to which it doesn't belong

File: local.py [hunk 7]
```diff
@@ -43,7 +48,7 @@
 
 flags.DEFINE_enum(
     'mode', 'classify',
-    ['embed', 'classify', 'prefetch', 'thresholds', 'predict', 'csv', 'import'],
+    ['embed', 'classify', 'prefetch', 'thresholds', 'csv', 'import', 'train'],
```
Comment (Tilana) [inline local.py]: what happened to the predict mode?

File: app/trainer.py [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+from datetime import datetime
+from distutils.dir_util import copy_tree
+from typing import Any, Callable, List, Set
+
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+from bert import tokenization
+from flask import Blueprint
+from models import class_based_attention
```
[the remark this answers is not in the document]
Comment (bdittes) [reply app/trainer.py]: It works for me an in github tests - did you ./run install?

File: README.md (hunk 1 above)
Comment (bdittes) [reply README.md]: That falls under ./run local --help - we didn't document model operations very well so far.

The trainer is also mostly intended to be used via HTTP by Uwazi - see test_e2e.py

File: app/trainer.py (hunk 6 above)
Comment (bdittes) [reply app/trainer.py]: vocab is about the model output (e.g. 'Torture'), not the BERT vocab. I added a comment.

File: local.py (hunk 7 above)
Comment (bdittes) [reply local.py]: predict was only used to update the predicted_labels in the internal classification_sample database. 'classify' is used to get predictions for a sequence.

Since it turned out that syncing Uwazi to the classification_sample db was hard (e.g. we'd need to detect deleted entities and stop using them for training), I decided the whole TC<->Uwazi communication would be easier by removing the need for the classification_sample db. It can still be used for ./run local --mode=train etc., but Uwazi now sends the whole training data with the train HTTP request and uses the /classify endpoint for predictions, so there was no need for the 'update predictions' task anymore and I deleted it.

**Comment by bdittes** [comment]
PTAL

File: app/trainer.py (hunk 6 above)
Comment (Tilana) [reply app/trainer.py]: let's please change vocab to labels. In ML these two have very different meaning

File: README.md (hunk 1 above)
Comment (Tilana) [reply README.md]: sorry, my fault.

File: app/trainer.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Tilana) [reply app/trainer.py]: sorry my fault

File: app/trainer.py (hunk 6 above)
Comment (bdittes) [reply app/trainer.py]: Changed everywhere
